src/Features/Core/Portable/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,337 @@
+                        }
+
+                        // Report IDE0051 or IDE0052 based on whether the underlying member has any Write/WritableRef/NonReadWriteRef references or not.
+                        var rule = !valueUsageInfo.ContainsWriteOrWritableRef() && !valueUsageInfo.ContainsNonReadWriteRef() && !symbolsReferencedInDocComments.Contains(member)
+                            ? _removeUnusedMembersRule
+                            : _removeUnreadMembersRule;
+                        var effectiveSeverity = rule.GetEffectiveSeverity(symbolEndContext.Compilation.Options);
+
+                        var diagnostic = DiagnosticHelper.Create(
+                            rule,
+                            member.Locations[0],
+                            effectiveSeverity,
+                            additionalLocations: member.Locations,
+                            properties: null,
+                            member.ContainingType.Name,
+                            member.Name);
+                        symbolEndContext.ReportDiagnostic(diagnostic);
+                    }
+                }
+
+                bool HasSyntaxErrors()
+                {
+                    foreach (var tree in symbolEndContext.Symbol.Locations.Select(l => l.SourceTree))
+                    {
+                        if (tree.GetDiagnostics(symbolEndContext.CancellationToken).Any(d => d.Severity == DiagnosticSeverity.Error))
+                        {
+                            return true;
+                        }
+                    }
+
+                    return false;
+                }
+
+                ImmutableHashSet<ISymbol> GetCandidateSymbolsReferencedInDocComments()
+                {
+                    var builder = ImmutableHashSet.CreateBuilder<ISymbol>();
+                    foreach (var root in symbolEndContext.Symbol.Locations.Select(l => l.SourceTree.GetRoot(symbolEndContext.CancellationToken)))
+                    {
+                        SemanticModel lazyModel = null;
+                        foreach (var node in root.DescendantNodes(descendIntoTrivia: true)
+                                                 .OfType<TDocumentationCommentTriviaSyntax>()
+                                                 .SelectMany(n => n.DescendantNodes().OfType<TIdentifierNameSyntax>()))
+                        {
+                            lazyModel = lazyModel ?? symbolEndContext.Compilation.GetSemanticModel(root.SyntaxTree);
+                            var symbol = lazyModel.GetSymbolInfo(node, symbolEndContext.CancellationToken).Symbol;
+                            if (symbol != null && IsCandidateSymbol(symbol))
+                            {
+                                builder.Add(symbol);
+                            }
+                        }
+                    }
+
+                    return builder.ToImmutable();
```

Comment:
note: this presumes that all the doc-comemtn references will be in the trees containing the type you are examining.  That's not necessarily true.  That said... i'm willing to give this a pass if we open a bug to track this being incorrect.

Reply:
Ah, I didn't know that doc comments can refer to private members of unrelated types :(.  I don't think we can get this right in the analyzer unless we have a compilation start/end action, which I would like to avoid. One thing we can possibly do is bail out from the code fix for this case, which might be acceptable given that this is a pretty corner-case scenario. I will file a tracking bug here.

---
In reply to: [213418852](https://github.com/dotnet/roslyn/pull/29511#discussion_r213418852) [](ancestors = 213418852)

Reply:
I'm find with that.  It seems sufficiently corner-casey to me :)

Reply:
Added a tracking comment at the end of [https://github.com/dotnet/roslyn/issues/29519#issue-354013140](https://github.com/dotnet/roslyn/issues/29519#issue-354013140)

---
In reply to: [213465207](https://github.com/dotnet/roslyn/pull/29511#discussion_r213465207) [](ancestors = 213465207)